subsys/bluetooth/gatt_dm.c

```diff
@@ -648,6 +648,7 @@ int bt_gatt_dm_start(struct bt_conn *conn,
 	if (err) {
 		LOG_ERR("Discover failed, error: %d.", err);
 		atomic_clear_bit(dm->state_flags, STATE_ATTRS_LOCKED);
+		discovery_complete_error(dm, err);
```

Comment:
If you want to call here this function, one line above makes more sense. Also, the current behavior of `bt_gatt_dm_start() is that if it returns an error, then no callback is executed, so you have to make sure that `error_found` is not call in this case.

And, I am also curious when this memory leak can happen. Can you give more details about reproduction scenario?